src/sbvr-api/permissions.ts

```diff
@@ -430,17 +481,117 @@ const generateConstrainedAbstractSql = (
 	const collapsedPermissionFilters = collapsePermissionFilters(
 		permissionFilters,
 	);
-	_.set(odata, ['tree', 'options', '$filter'], collapsedPermissionFilters);
 
-	const translated = uriParser.translateUri({
-		method: 'GET',
-		resourceName,
+	return collapsedPermissionFilters;
+};
+
+const constrainedPermissionError = new PermissionError();
+const generateConstrainedAbstractSql = (
+	permissionsLookup: PermissionLookup,
+	actionList: PermissionCheck,
+	vocabulary: string,
+	resourceName: string,
+) => {
+	const abstractSQLModel = sbvrUtils.getAbstractSqlModel({
 		vocabulary,
-		odataBinds: odata.binds,
-		odataQuery: odata.tree,
-		values: {},
 	});
-	const abstractSqlQuery = _.clone(translated.abstractSqlQuery!);
+	const odata = memoizedParseOdata(`/${resourceName}`);
+
+	const collapsedPermissionFilters = buildODataPermission(
+		permissionsLookup,
+		actionList,
+		vocabulary,
+		resourceName,
+		odata,
+	);
+
+	_.set(odata, ['tree', 'options', '$filter'], collapsedPermissionFilters);
+
+	const lambdaAlias = randomstring.generate(20);
+	let inc = 0;
+
+	// We need to trace the processed resources, to be able to break
+	// permissions circles.
+	const canAccessTrace: string[] = [resourceName];
+
+	const canAccessFunction: ResourceFunction = function(property: AnyObject) {
+		// remove method property so that we won't loop back here again at this point
+		delete property.method;
+
+		if (!this.defaultResource) {
+			throw new Error(`No resource selected in AST.`);
+		}
+
+		const targetResource = this.NavigateResources(
+			this.defaultResource,
+			property.name,
+		);
+
+		const targetResourceName = sqlNameToODataName(targetResource.resource.name);
+
+		if (canAccessTrace.includes(targetResourceName)) {
+			// return a boolen subquery
+			return [
+				'Not',
+				[
+					'Exists',
+					['SelectQuery', ['Select', [['Alias', ['Number', 1], 'id']]]],
+				],
+			];
+		}
+
+		const parentOdata = memoizedParseOdata(`/${targetResourceName}`);
+
+		const collapsedParentPermissionFilters = buildODataPermission(
+			permissionsLookup,
+			actionList,
+			vocabulary,
+			targetResourceName,
+			parentOdata,
+		);
+
+		if (collapsedParentPermissionFilters === false) {
+			// We reuse a constant permission error here as it will be cached, and
+			// using a single error instance can drastically reduce the memory used
+			throw constrainedPermissionError;
+		}
+
+		const lambdaId = `${lambdaAlias}+${inc}`;
+		inc = inc + 1;
+		rewriteSubPermissionBindings(
+			collapsedParentPermissionFilters,
+			this.bindVarsLength + this.extraBindVars.length,
+		);
+		convertToLambda(collapsedParentPermissionFilters, lambdaId);
+		property.lambda = {
+			method: 'any',
+			identifier: lambdaId,
+			expression: collapsedParentPermissionFilters,
+		};
+
+		this.extraBindVars.push(...parentOdata.binds);
+
+		canAccessTrace.push(targetResourceName);
+		const value = this.Property(property);
+		canAccessTrace.pop();
+		return value;
```

Comment:
Using a try/finally would make this safer from edge cases/future changes:
```suggestion
		canAccessTrace.push(targetResourceName);
		try {
			return this.Property(property);
		} finally {
			canAccessTrace.pop();
		}
```